src/Microsoft.AspNetCore.Hosting/Internal/HostingApplicationDiagnostics.cs

```diff
@@ -0,0 +1,276 @@
+        private const string DiagnosticsUnhandledExceptionKey = "Microsoft.AspNetCore.Hosting.UnhandledException";
+
+        private const string RequestIdHeaderName = "Request-Id";
+        private const string CorrelationContextHeaderName = "Correlation-Context";
+
+        private readonly DiagnosticListener _diagnosticListener;
+        private readonly ILogger _logger;
+
+        public HostingApplicationDiagnostics(ILogger logger, DiagnosticListener diagnosticListener)
+        {
+            _logger = logger;
+            _diagnosticListener = diagnosticListener;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void BeginRequest(HttpContext httpContext, ref HostingApplication.Context context)
+        {
+            // These enabled checks are virtual dispatch and used twice and so cache to locals
+            var diagnosticsEnabled = _diagnosticListener.IsEnabled();
+            var legacyDiagnosticsEnabled = diagnosticsEnabled && _diagnosticListener.IsEnabled(DiagnosticsBeginRequestKey);
+            var loggingEnabled = _logger.IsEnabled(LogLevel.Information);
+            var eventLogEnabled = HostingEventSource.Log.IsEnabled();
+
+            if (eventLogEnabled)
+            {
+                context.EventLogEnabled = true;
+                // To keep the hot path short we defer logging in this function to non-inlines
+                RecordRequestStartEventLog(httpContext);
+            }
+
+            // Only make call GetTimestamp if its value will be used, i.e. of the listenters is enabled
+            var startTimestamp = (legacyDiagnosticsEnabled || loggingEnabled) ? Stopwatch.GetTimestamp() : 0;
+
+            // Scope may be relevant for a different level of logging, so we always create it
+            // see: https://github.com/aspnet/Hosting/pull/944
+            context.Scope = HostingLoggerExtensions.RequestScope(_logger, httpContext);
+
+            if (diagnosticsEnabled && _diagnosticListener.IsEnabled(ActivityName, httpContext))
+            {
+                context.Activity = StartActivity(httpContext);
+            }
+            if (loggingEnabled)
+            {
+                // Non-inline
+                LogRequestStarting(httpContext);
+            }
+            if (legacyDiagnosticsEnabled)
+            {
+                // Non-inline
+                RecordBeginRequestDiagnostics(httpContext, startTimestamp);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void RequestEnd(HttpContext httpContext, Exception exception, HostingApplication.Context context)
+        {
+            // Local cache items resolved multiple items, in order of use so they are primed in cpu pipeline when used
+            var startTimestamp = context.StartTimestamp;
+            var diagnosticsEnabled = _diagnosticListener.IsEnabled();
```

Comment:
Same idea as for RequestStart.   We should have a minimum of work (tests) when things are off.  Thus 3 tests.   Will look a lot like the RequestStart case.